File: easyapi_java/src/main/java/com/hlab/easyapi_java/easyapi/main/api/EasyApiRx.java
package com.hlab.easyapi_java.easyapi.main.api;

import android.support.annotation.NonNull;
import com.hlab.easyapi_java.R;
import com.hlab.easyapi_java.easyapi.util.NetworkUtils;
import com.hlab.easyapi_java.easyapi.util.ResponseHandler;
import io.reactivex.Observable;
import io.reactivex.ObservableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by H.T. on 05/01/18.
 */

public class EasyApiRx<T> extends EasyApiBase implements EasyApiCall<T> {

  private Observable<T> call;
  private DisposableObserver<T> disposable;

  EasyApiRx(Config config) {
    this.config = config;
  }

  @Override
  public EasyApiRx<T> initCall(@NonNull Observable<T> call) {
    this.call = call;
    return this;
  }

  @Override
  public void execute(boolean showErrorMessages, @NonNull ResponseHandler<T> responseHandler) {
    DisposableObserver<T> disposable = getDisposableObserver(responseHandler, showErrorMessages);

    Observable<T> source = getSource();

    if (source != null) {
      showProgress();

      source.subscribe(disposable);
      this.disposable = disposable;
    }
  }

  @Override
  public void executeWith(@NonNull DisposableObserver<T> observer) {
    Observable<T> source = getSource();

    if (source != null) {
      source.subscribe(observer);
    }
  }

  @Override
  public void executeWith(@NonNull Consumer<T> successConsumer,
      @NonNull Consumer<Throwable> failureConsumer) {
    Observable<T> source = getSource();

    if (source != null) {
      source.subscribe(successConsumer, failureConsumer);
    }
  }

  private Observable<T> getSource() {
    return NetworkUtils.isNetworkAvailable(config.getmContext()).filter(available -> {
      if (!available) {
        showNoInternet();
      }
      return available;
    }).switchMap(available -> {
      if (available) {
        return call.compose(applyObservableAsync());
      }
      return Observable.empty();
    });
  }

  @NonNull
  private DisposableObserver<T> getDisposableObserver(ResponseHandler<T> responseHandler,
      boolean showErrorMessages) {
    return new DisposableObserver<T>() {
      @Override
      public void onNext(T responseBody) {
        try {
          if (responseBody != null) {
            responseHandler.onResponse(responseBody, true);
          } else {
            responseHandler.onResponse(null, false);
            if (showErrorMessages) {
              showError(config.getmContext().getString(R.string.alert_message_error));
            }
          }
        } catch (Exception e) {
          e.printStackTrace();
          responseHandler.onResponse(null, false);
          if (showErrorMessages) {
            showError(config.getmContext().getString(R.string.alert_message_error));
          }
        }
      }

      @Override
      public void onError(Throwable e) {
        responseHandler.onResponse(null, false);
        if (showErrorMessages) {
          showError(NetworkUtils.handleErrorResponse(e));
        }
      }

      @Override
      public void onComplete() {
        hideProgress();
      }
    };
  }

  @Override
  public void dispose() {
    if (disposable != null && !disposable.isDisposed()) {
      disposable.dispose();
    }
  }

  /**
   * Generate reusable observable source object.
   *
   * @return ObservableTransformer
   */
  private <F> ObservableTransformer<F, F> applyObservableAsync() {
    return observable -> observable.subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread());
  }
}

File: easyapi_kotlin/src/main/java/com/hlab/easyapi_kotlin/easyapi/main/api/EasyApiRx.kt
package com.hlab.easyapi_kotlin.easyapi.main.api

import com.hlab.easyapi_kotlin.R
import com.hlab.easyapi_kotlin.easyapi.util.NetworkUtils
import io.reactivex.Observable
import io.reactivex.ObservableTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers

/**
 * Created by H.T. on 05/01/18.
 */

class EasyApiRx<T> internal constructor(config: Config) : EasyApiBase(), EasyApiCall<T> {

  private var call: Observable<T>? = null
  private var disposable: DisposableObserver<T>? = null

  private val source: Observable<T>?
    get() = NetworkUtils.isNetworkAvailable(config.getmContext())
        .filter { available ->
          if (!available) {
            showNoInternet()
          }
          available
        }
        .switchMap { available ->
          if (available!!) {
            call?.compose<T>(applyObservableAsync<T>())
          } else
            Observable.empty<T>()
        }

  init {
    this.config = config
  }

  override fun initCall(call: Observable<T>): EasyApiRx<T>? {
    this.call = call
    return this
  }

  override fun execute(
    showErrorMessages: Boolean,
    responseHandler: (response: T?, isError: Boolean) -> Unit/*responseHandler: ResponseHandler<T>*/
  ) {
    val disposable = getDisposableObserver(responseHandler, showErrorMessages)

    val source = source

    if (source != null) {
      showProgress()

      source.subscribe(disposable)
      this.disposable = disposable
    }
  }

  override fun executeWith(observer: DisposableObserver<T>) {
    val source = source

    source?.subscribe(observer)
  }

  override fun executeWith(
    successConsumer: Consumer<T>,
    failureConsumer: Consumer<Throwable>
  ) {
    val source = source

    source?.subscribe(successConsumer, failureConsumer)
  }

  private fun getDisposableObserver(
    responseHandler: (response: T?, isError: Boolean) -> Unit/*responseHandler: ResponseHandler<T>*/,
    showErrorMessages: Boolean
  ): DisposableObserver<T> {
    return object : DisposableObserver<T>() {
      override fun onNext(responseBody: T?) {
        try {
          if (responseBody != null) {
            responseHandler.invoke(responseBody, true)
          } else {
            responseHandler.invoke(null, false)
            if (showErrorMessages) {
              showError(config.getmContext().getString(R.string.alert_message_error))
            }
          }
        } catch (e: Exception) {
          e.printStackTrace()
          responseHandler.invoke(null, false)
          if (showErrorMessages) {
            showError(config.getmContext().getString(R.string.alert_message_error))
          }
        }

      }

      override fun onError(e: Throwable) {
        responseHandler.invoke(null, false)
        if (showErrorMessages) {
          showError(NetworkUtils.handleErrorResponse(e))
        }
      }

      override fun onComplete() {
        hideProgress()
      }
    }
  }

  override fun dispose() {
    disposable?.let {
      if (!it.isDisposed)
        it.dispose()
    }
  }

  /**
   * Generate reusable observable source object.
   *
   * @return ObservableTransformer
   */
  private fun <F> applyObservableAsync(): ObservableTransformer<F, F> {
    return ObservableTransformer { observable ->
      observable.subscribeOn(Schedulers.newThread())
          .observeOn(AndroidSchedulers.mainThread())
    }
  }
}
